docs/design/jet/023-jobupload.md

```diff
@@ -0,0 +1,164 @@
+| Support Engineer               | _TBD_                                            |
+| Technical Reviewers            | _Frantisek Hartman_                              |
+| Simulator or Soak Test PR(s)   | _TBD_                                            |
+
+### Background
+#### Description
+
+- Currently, the job upload can only be performed by hz-cli command. This command requires a JVM on the host machine. The purpose of this feature is to enable non-java clients to be able to upload jobs to Jet server 
+
+
+#### Terminology
+
+| Term                   | Definition                                                              |
+|------------------------|-------------------------------------------------------------------------|
+| Client Binary Protocol | The binary messages used between Hazelcast Clients and Hazelcast Server |
+
+#### Actors and Scenarios
+
+Any language that implements Client Binary Protocol
+
+### Functional Design
+#### Summary of Functionality
+
+Provide a list of functions user(s) can perform.
+- Upload a jar file and pass the same parameters that can be used with hz-cli
+
+#### Additional Functional Design Topics
+
+None
+
+##### Notes/Questions/Issues
+
+- The new functionality requires that all resources accessed by the uploaded job to be available on the server. For example a text file that is used to populate an IMap needs to be available on the server or within the uploaded jar. Because a server side job can not access client side resources.  
+- The client should send multi parts in a sequential manner. Out of order messages are not handled since it would require more resources on the server.
+  
+
+### User Interaction
+#### API design and/or Prototypes
+
+```java
+HazelcastInstance client = ...;
+JetService jetService = client.getJet();
+List<String> jobParameters = emptyList();
+
+String job1 = "job1";
+//If there is an error, throws JetException
+jetService.submitJobJar(getJarPath(),
+        null,
+        job1,
+        null,
+        jobParameters);
+```
+
+#### Client Related Changes
+A new method has been added to JetService interface. Any client that implements this interface, has to implement this method
+```java
+void submitJobJar(@Nonnull Path jarPath, String snapshotName, String jobName, String mainClass, List<String> jobParameters);
+```
+
+### Technical Design
+
+The client protocol needs to support job uploading. So that non-java clients can also upload jet jobs.
+For this purpose two new messages are added to client protocol
+
+1. uploadJobMetaData
+2. uploadJobMultipart
+
+**1. uploadJobMetaData Message**
+
+The upload process starts with uploadJobMetaData. This message contains
+
+| Term          | Type        | Definition                                                |
+|---------------|-------------|-----------------------------------------------------------|
+| sessionId     | UUID        | The UUID. This field associates all messages in a session |
```

Comment:
Is this used as the `jobID`? Otherwise is it possible to return the `jobID` in the response of this message?

Reply:
This field is not the jobID. It is just used for session management during job upload . As far as I understand there is not way to return the jobID from the upload process. But the client can query jobs to get the IDs